[722] Import vacancies from the production API to other environments

robbiepaul opened the pull request:
## Trello card URL:

https://trello.com/c/hjkz9DrM/722-ensure-environment-data-parity-between-production-and-staging-testing

## Changes in this PR:

The aim of this PR is to import vacancies from production to other environments so that we have meaningful data.

- Rake task to fetch the latest 50 vacancies from the production API and queue a job up for each
- The job then maps the `JobPosting` values to a `Vacancy` and attempts to make it valid before saving

## Questions
- Do we want to make this task run on a schedule?

## Next steps:

- [x] Terraform deployment required?

- [ ] New development configuration to be shared?

robbiepaul commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Ensure environment data parity between production and staging/testing](https://trello.com/c/hjkz9DrM/722-ensure-environment-data-parity-between-production-and-staging-testing)

pezholio commented on lib/tasks/seed_vacancies_from_api.rake:
```diff
@@ -0,0 +1,16 @@
+  desc 'Seed from the Teaching Vacancies API'
+  namespace :seed_from_api do
+    task vacancies: :environment do
+      next if Rails.env.production?
```
💯

pezholio commented on lib/tasks/seed_vacancies_from_api.rake:
```diff
@@ -0,0 +1,16 @@
+      next if Rails.env.production?
+
+      Rails.logger.debug("Seeding vacancies from Teaching Vacancies API in #{Rails.env}")
+      job_postings = TeachingVacancies::API.new.jobs(limit: 50)
```
Do we need to specify the limit here, as it's already 50 by default? I'd probably leave the limit here, and set a default limit to be lower, just to be explicit

robbiepaul replied:
Makes sense, I wondered whether to have a limit at all but I've set the default limit to be lower 👍

pezholio commented on lib/teaching_vacancies_api.rb:
```diff
@@ -0,0 +1,19 @@
+      json = JSON.parse(response.body)
+      json['data'].take(limit)
+    rescue HTTParty::ResponseError => e
+      Rails.logger.warn("Teaching Vacancies API response error: #{e.message}")
```
We probably want to raise the error and handle it in the Rake task too, as it won't return an array and it won't immediately be clear what's happened

robbiepaul replied:
I've added a commit that moves the `rescue` to the rake task, thanks for pointing it out 👍

pezholio commented on spec/lib/tasks/seed_vacancies_from_api_rake_spec.rb:
```diff
@@ -0,0 +1,26 @@
+  end
+
+  context 'when in production' do
+    before { allow(Rails.env).to receive(:production?).and_return(true) }
```
💯 💯

robbiepaul commented:
@tahb To give us some more confidence I've added two more guards, one that checks for the `FEATURE_IMPORT_VACANCIES` env var is set to true and the other checks the `current_database` name is in a whitelist.

I've also added logging of the imported vacancy ID incase we need to delete any.

tahb commented on db/schema.rb:
```diff
@@ -10,7 +10,7 @@
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema.define(version: 2019_03_18_181006) do
+ActiveRecord::Schema.define(version: 2019_03_21_180503) do
```
Oh oh I think schema has been accidentally committed here

robbiepaul replied:
Ooops, I've uncommited (is that a word?) that file. Specs are passing again now too

tahb commented:
Tested on staging, it adds 30 listings and logs the new ID to the logs!

20 failures due to certain fields being blank. This seems odd since everything exposed by the API must have gone through the creation validation right?

```
Mar 27 10:26:19 Edge ecs-tvs2_edge_worker-101-tvs2edgeweb-f0a280c08299e7b3d501:  Failed to save vacancy from JobPosting: {:minimum_salary=>[], :maximum_salary=>[], :education=>["can't be blank"], :qualifications=>["can't be blank"]}
Mar 27 10:26:19 Edge ecs-tvs2_edge_worker-101-tvs2edgeweb-f0a280c08299e7b3d501:  Failed to save vacancy from JobPosting: {:minimum_salary=>[], :maximum_salary=>[], :education=>["can't be blank"], :qualifications=>["can't be blank"]}
Mar 27 10:26:20 Edge ecs-tvs2_edge_worker-101-tvs2edgeweb-f0a280c08299e7b3d501:  Failed to save vacancy from JobPosting: {:minimum_salary=>[], :maximum_salary=>[], :experience=>["can't be blank"], :education=>["can't be blank"], :qualifications=>["can't be blank"]}
Mar 27 10:26:20 Edge ecs-tvs2_edge_worker-101-tvs2edgeweb-f0a280c08299e7b3d501:  Failed to save vacancy from JobPosting: {:minimum_salary=>[], :maximum_salary=>[], :education=>["can't be blank"], :qualifications=>["can't be blank"]}
Mar 27 10:26:20 Edge ecs-tvs2_edge_worker-101-tvs2edgeweb-f0a280c08299e7b3d501:  Failed to save vacancy from JobPosting: {:minimum_salary=>[], :maximum_salary=>[], :education=>["can't be blank"], :qualifications=>["can't be blank"]}
Mar 27 10:26:20 Edge ecs-tvs2_edge_worker-101-tvs2edgeweb-f0a280c08299e7b3d501:  Failed to save vacancy from JobPosting: {:minimum_salary=>[], :maximum_salary=>[], :experience=>["is too long (maximum is 1000 characters)"]}
Mar 27 10:26:21 Edge ecs-tvs2_edge_worker-101-tvs2edgeweb-f0a280c08299e7b3d501:  Failed to save vacancy from JobPosting: {:minimum_salary=>[], :maximum_salary=>[], :experience=>["can't be blank"], :education=>["can't be blank"], :qualifications=>["can't be blank"]}
Mar 27 10:26:21 Edge ecs-tvs2_edge_worker-101-tvs2edgeweb-acf4c7929fa59a9e9201:  Failed to save vacancy from JobPosting: {:minimum_salary=>[], :maximum_salary=>[], :experience=>["can't be blank"], :education=>["can't be blank"], :qualifications=>["can't be blank"]}
Mar 27 10:26:21 Edge ecs-tvs2_edge_worker-101-tvs2edgeweb-acf4c7929fa59a9e9201:  Failed to save vacancy from JobPosting: {:minimum_salary=>[], :maximum_salary=>[], :experience=>["can't be blank"], :education=>["can't be blank"], :qualifications=>["can't be blank"]}
Mar 27 10:26:21 Edge ecs-tvs2_edge_worker-101-tvs2edgeweb-f0a280c08299e7b3d501:  Failed to save vacancy from JobPosting: {:minimum_salary=>[], :maximum_salary=>[], :experience=>["can't be blank"], :education=>["can't be blank"], :qualifications=>["can't be blank"]}
Mar 27 10:26:21 Edge ecs-tvs2_edge_worker-101-tvs2edgeweb-acf4c7929fa59a9e9201:  Failed to save vacancy from JobPosting: {:minimum_salary=>[], :maximum_salary=>[], :experience=>["can't be blank"], :education=>["can't be blank"], :qualifications=>["can't be blank"]}
Mar 27 10:26:21 Edge ecs-tvs2_edge_worker-101-tvs2edgeweb-f0a280c08299e7b3d501:  Failed to save vacancy from JobPosting: {:minimum_salary=>[], :maximum_salary=>[], :experience=>["can't be blank"], :education=>["can't be blank"], :qualifications=>["can't be blank"]}
Mar 27 10:26:21 Edge ecs-tvs2_edge_worker-101-tvs2edgeweb-acf4c7929fa59a9e9201:  Failed to save vacancy from JobPosting: {:minimum_salary=>[], :maximum_salary=>[], :experience=>["can't be blank"], :education=>["can't be blank"], :qualifications=>["can't be blank"]}
Mar 27 10:26:21 Edge ecs-tvs2_edge_worker-101-tvs2edgeweb-f0a280c08299e7b3d501:  Failed to save vacancy from JobPosting: {:minimum_salary=>[], :maximum_salary=>[], :experience=>["can't be blank"], :education=>["can't be blank"], :qualifications=>["can't be blank"]}
Mar 27 10:26:21 Edge ecs-tvs2_edge_worker-101-tvs2edgeweb-9cb2a1be808382adb201:  Failed to save vacancy from JobPosting: {:minimum_salary=>[], :maximum_salary=>[], :experience=>["can't be blank"], :education=>["can't be blank"], :qualifications=>["can't be blank"]}
Mar 27 10:26:21 Edge ecs-tvs2_edge_worker-101-tvs2edgeweb-9cb2a1be808382adb201:  Failed to save vacancy from JobPosting: {:minimum_salary=>[], :maximum_salary=>[], :experience=>["can't be blank"], :education=>["can't be blank"], :qualifications=>["can't be blank"]}
Mar 27 10:26:21 Edge ecs-tvs2_edge_worker-101-tvs2edgeweb-9cb2a1be808382adb201:  Failed to save vacancy from JobPosting: {:minimum_salary=>[], :maximum_salary=>[], :education=>["can't be blank"], :qualifications=>["can't be blank"]}
Mar 27 10:26:22 Edge ecs-tvs2_edge_worker-101-tvs2edgeweb-9cb2a1be808382adb201:  Failed to save vacancy from JobPosting: {:minimum_salary=>[], :maximum_salary=>[], :experience=>["can't be blank"], :education=>["can't be blank"], :qualifications=>["can't be blank"]}
Mar 27 10:26:22 Edge ecs-tvs2_edge_worker-101-tvs2edgeweb-acf4c7929fa59a9e9201:  Failed to save vacancy from JobPosting: {:minimum_salary=>[], :maximum_salary=>[], :experience=>["is too long (maximum is 1000 characters)"]}
Mar 27 10:26:22 Edge ecs-tvs2_edge_worker-101-tvs2edgeweb-acf4c7929fa59a9e9201:  Failed to save vacancy from JobPosting: {:minimum_salary=>[], :maximum_salary=>[], :experience=>["can't be blank"], :education=>["can't be blank"], :qualifications=>["can't be blank"]}
```

robbiepaul commented:
@tahb Yeah, it looks like older vacancies didn't have the same validation rules. If you look on https://teaching-vacancies.service.gov.uk/api/v1/jobs.json you'll see quite a few `"qualifications": null,` etc. It looks like these vacancies are from early 2018.

That's why I've changed the ordering of the API results in b733bb9ad18e52bb1632f6a9d6765875cc898968 so that when this does make it to production we'll just be importing the freshest vacancies.

tahb commented:
@robbiepaul Nice one 👍

tahb commented:
I've added the new config to staging, edge and testing tfvars in preparation.
